testutil/checkers.go

```diff
@@ -225,3 +225,65 @@ func (c *deepContainsChecker) Check(params []interface{}, names []string) (resul
 		return false, fmt.Sprintf("%T is not a supported container", container)
 	}
 }
+
+type syscallsEqualChecker struct {
+	*check.CheckerInfo
+}
+
+// SyscallsEqual checks that one sequence of system calls is equal to another.
+var SyscallsEqual check.Checker = &syscallsEqualChecker{
+	CheckerInfo: &check.CheckerInfo{Name: "SyscallsEqual", Params: []string{"actualList", "expectedList"}},
+}
+
+func (c *syscallsEqualChecker) Check(params []interface{}, names []string) (result bool, error string) {
+	actualList, ok := params[0].([]CallResultError)
+	if !ok {
+		return false, "left-hand-side argument must be of type []CallResultError"
+	}
+	expectedList, ok := params[1].([]CallResultError)
+	if !ok {
+		return false, "right-hand-side argument must be of type []CallResultError"
+	}
+
+	for i, actual := range actualList {
+		if i >= len(expectedList) {
+			return false, fmt.Sprintf("system call #%d %#q unexpectedly present, got %d system call(s) but expected only %d",
+				i, actual.C, len(actualList), len(expectedList))
+		}
+		expected := expectedList[i]
+		if actual.C != expected.C {
+			return false, fmt.Sprintf("system call #%d differs in operation, actual %#q, expected %#q", i, actual.C, expected.C)
+		}
+		if !reflect.DeepEqual(actual, expected) {
+			switch {
+			case actual.E == nil && expected.E == nil && !reflect.DeepEqual(actual.R, expected.R):
+				// The call succeeded but not like we expected.
+				return false, fmt.Sprintf("system call #%d %#q differs in result, actual: %#v, expected: %#v",
+					i, actual.C, actual.R, expected.R)
+			case actual.E != nil && expected.E != nil && !reflect.DeepEqual(actual.E, expected.E):
+				// The call failed but not like we expected.
+				return false, fmt.Sprintf("system call #%d %#q differs in error, actual: %s, expected: %s",
+					i, actual.C, actual.E, expected.E)
+			case actual.E != nil && expected.E == nil && expected.R == nil:
+				// The call failed but we expected it to succeed.
+				return false, fmt.Sprintf("system call #%d %#q unexpectedly failed, actual error: %s", i, actual.C, actual.E)
+			case actual.E != nil && expected.E == nil && expected.R != nil:
+				// The call failed but we expected it to succeed with some result.
+				return false, fmt.Sprintf("system call #%d %#q unexpectedly failed, actual error: %s, expected result: %#v", i, actual.C, actual.E, expected.R)
+			case actual.E == nil && expected.E != nil && actual.R == nil:
+				// The call succeeded with some result but we expected it to fail.
+				return false, fmt.Sprintf("system call #%d %#q unexpectedly succeeded, expected error: %s", i, actual.C, expected.E)
```

Comment:
Note, this also combines the case where actual.R is nil and actual.R is !nil, but I think that is ok because we want to tell the developer how it should've failed.